File: paintHouseII.py
#====O(nK) time solution
class Solution(object):
    def minCostII(self, costs):
        """
        :type costs: List[List[int]]
        :rtype: int
        """
        if len(costs) == 0: return 0
        m, k = len(costs), len(costs[0])
        #started from second row, reuse the costs array to store our record
        for row in xrange(1, m):
            # find min and secondMin in the cost row before
            minColor, secColor = -1, -1
            for color in xrange(k):
                cost = costs[row-1][color]
                if (minColor == -1 or cost < costs[row-1][minColor]):
                    secColor = minColor
                    minColor = color
                elif (secColor == -1 or cost < costs[row-1][secColor]):
                    secColor = color
            
            #assign cost table in current row
            for color in xrange(k):
                if color == minColor:
                    costs[row][color] += costs[row-1][secColor]
                else:
                    costs[row][color] += costs[row-1][minColor]
        return min(costs[m-1])

File: paintHouseII.kt
class Solution {
    fun minCostII(costs: Array<IntArray>): Int {
        // var colors = Array(costs.size){ Array(2){IntArray(2)} } // for each row, save the min cost and its color number, and second min cost and its color number
        // var colors = Array(2){IntArray(2)} 
        // [0][0], [0][1] stands for        min cost and its color id
        // [1][0], [1][1] stands for second min cost and its color id
        var prevMinCost = 0
        var prevMinColor = -1
        var prevSecMinCost = 0

        costs.forEach{cost ->
            val pq = PriorityQueue<IntArray>(compareBy {-it[0]}) // we are using a max heap here
            cost.forEachIndexed{idx,  price  ->
                if (idx == prevMinColor) // if the color collide with previous house, use the second min cost, otherwise use the min cost strightly
                    pq.add(intArrayOf(price + prevSecMinCost, idx))
                else pq.add(intArrayOf(price + prevMinCost, idx))
                if ( pq.size > 2) pq.poll()
            }
            val secondMinPair = pq.poll() // (cost, idx) color with second min cost on i house
            val minPair = pq.poll() // (cost, idx) color with min cost on i house
            prevMinCost = minPair[0]
            prevMinColor = minPair[1]
            prevSecMinCost = secondMinPair[0]
        }
        return prevMinCost
    }
}
